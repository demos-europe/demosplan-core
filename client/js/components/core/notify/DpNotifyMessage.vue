<license>
  (c) 2010-present DEMOS E-Partizipation GmbH.

  This file is part of the package demosplan,
  for more information see the license file.

  All rights reserved
</license>

<template>
  <div
    :class="prefixClass('c-notify__message ' + messageClass)">
    <i
      :class="prefixClass('c-notify__closer fa fa-times-circle cursor--pointer')"
      aria-hidden="true"
      @click.stop.prevent="hide" />

    <div :class="prefixClass('cf')">
      <i
        :class="prefixClass('fa u-mt-0_125 u-mr-0_25 float--left ' + messageIcon)" />
      <div :class="prefixClass('u-ml')">
        {{ message.text }}
        <a
          v-if="message.linkUrl"
          :class="prefixClass('c-notify__link u-mt-0_25')"
          :href="message.linkUrl"
          data-cy="messageLink">
          {{ message.linkText || message.linkUrl }}
        </a>
      </div>
    </div>
  </div>
</template>

<script>
import { prefixClassMixin } from 'demosplan-ui/mixins'

export default {
  name: 'DpNotifyMessage',

  mixins: [prefixClassMixin],

  props: {
    /**
     * A single notification message object containing type, text(, linkUrl, linkText)
     */
    message: {
      type: Object,
      required: true
    },
    hideTimer: {
      type: Number,
      required: false,
      default: 20000
    }
  },

  data () {
    return {
      messageId: '',
      icons: {
        info: 'fa-info-circle',
        warning: 'fa-exclamation-triangle',
        error: 'fa-exclamation-circle',
        confirm: 'fa-check-circle',
        dev: 'fa-info-circle'
      }
    }
  },

  computed: {
    messageClass () {
      return 'c-notify__message--' + this.message.type
    },

    messageIcon () {
      return this.icons[this.message.type]
    }
  },

  methods: {
    hide () {
      this.$emit('dp-notify-remove', this.message)
    }
  },

  mounted () {
    if (this.message.type !== 'error') {
      setTimeout(() => {
        this.hide()
      }, this.hideTimer)
    }
  }
}
</script>
