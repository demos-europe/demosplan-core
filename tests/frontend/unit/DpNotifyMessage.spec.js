/**
 * (c) 2010-present DEMOS E-Partizipation GmbH.
 *
 * This file is part of the package demosplan,
 * for more information see the license file.
 *
 * All rights reserved
 */

import DpNotifyMessage from '@DpJs/components/core/notify/DpNotifyMessage'
import { mount } from '@vue/test-utils'

describe('DpNotifyMessage', () => {
  it('is named DpNotifyMessage', () => {
    expect(DpNotifyMessage.hasOwnProperty('name')).toBe(true)
    expect(DpNotifyMessage.name).toBe('DpNotifyMessage')
  })

  it('chooses the message class based on it\'s type', () => {
    const validCombinations = [
      {
        type: 'confirm',
        clazz: 'c-notify__message--confirm'
      },
      {
        type: 'info',
        clazz: 'c-notify__message--info'
      },
      {
        type: 'warning',
        clazz: 'c-notify__message--warning'
      }
    ]

    for (const val of validCombinations) {
      const wrapper = mount(DpNotifyMessage, {
        propsData: {
          message: {
            type: val.type
          }
        }
      })

      expect(wrapper.vm.messageClass).toBe(val.clazz)
    }
  })

  it('always renders with closemark', () => {
    let wrapper = mount(DpNotifyMessage, {
      propsData: {
        message: {
          type: 'error'
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    let closer = wrapper.find('i.c-notify__closer')
    expect(closer.element.tagName).toStrictEqual('I')

    wrapper = mount(DpNotifyMessage, {
      propsData: {
        message: {
          type: 'confirm'
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    closer = wrapper.find('i.c-notify__closer')
    expect(closer.element.tagName).toStrictEqual('I')
  })

  it('renders the message text', () => {
    const wrapper = mount(DpNotifyMessage, {
      propsData: {
        message: {
          type: 'confirm',
          text: 'MessageText'
        }
      }
    })

    expect(wrapper.find('.cf > .u-ml').text()).toBe('MessageText')
  })

  it('renders a link if link attributes are given', () => {
    const wrapper = mount(DpNotifyMessage, {
      propsData: {
        message: {
          type: 'confirm',
          text: 'MessageText',
          linkUrl: 'about:blank',
          linkText: 'LinkText'
        }
      }
    })

    expect(wrapper.find('.cf > .u-ml').text()).toBe('MessageText\n      \n        LinkText')
  })

  it('emits dp-notify-remove with it\'s message as payload once clicked', (done) => {
    const message = {
      type: 'confirm',
      text: 'MessageText'
    }

    const wrapper = mount(DpNotifyMessage, {
      propsData: {
        message: message
      }
    })

    wrapper.vm.$on('dp-notify-remove', (payload) => {
      expect(payload).toBe(message)
      done()
    })

    wrapper.find('.c-notify__closer').trigger('click')
  })

  it('emits dp-notify-remove after hide timeout idle', (done) => {
    const message = {
      type: 'confirm',
      text: 'MessageText'
    }

    const wrapper = mount(DpNotifyMessage, {
      propsData: {
        message: message,
        hideTimer: 25 // Make this timeout reasonably short to keep the test time low
      }
    })

    wrapper.vm.$on('dp-notify-remove', (payload) => {
      expect(payload).toBe(message)
      done()
    })
  })
})
