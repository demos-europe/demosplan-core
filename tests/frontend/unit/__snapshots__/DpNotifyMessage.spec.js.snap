// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DpNotifyMessage always renders with closemark 1`] = `
"<div class=\\"c-notify__message c-notify__message--error\\"><i aria-hidden=\\"true\\" class=\\"c-notify__closer fa fa-times-circle cursor--pointer\\"></i>
  <div class=\\"cf\\"><i class=\\"fa u-mt-0_125 u-mr-0_25 float--left fa-exclamation-circle\\"></i>
    <div class=\\"u-ml\\">

      <!---->
    </div>
  </div>
</div>"
`;

exports[`DpNotifyMessage always renders with closemark 2`] = `
"<div class=\\"c-notify__message c-notify__message--confirm\\"><i aria-hidden=\\"true\\" class=\\"c-notify__closer fa fa-times-circle cursor--pointer\\"></i>
  <div class=\\"cf\\"><i class=\\"fa u-mt-0_125 u-mr-0_25 float--left fa-check-circle\\"></i>
    <div class=\\"u-ml\\">

      <!---->
    </div>
  </div>
</div>"
`;
